Add support for hybrid environments (#12)
bottlerocket-os/bottlerocket-control-container: README.md

You'll need Docker 17.06.2 or later, for multi-stage build support.
Then run `make`!

## Connecting to AWS Systems Manager (SSM)

> Starting from v0.5.0, users have the option to pass in their own activations for SSM. This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.

Reviewer: Our convention has been to write markdown files with one sentence per line.  If there are no empty lines, they get formatted as a paragraph.  If you want a new paragraph, leave a blank line.  It just makes these docs a bit easier to read/edit.

> Starting from v0.5.0, users have the option to pass in their own activations for SSM. This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.

Reviewer: Should this say "pass in their own activation information"?

Author: I'm not sure actually. The AWS documentation refers to them as Hybrid Activations, but both I think could be correct.

> Users can add their [own activations](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-managed-instance-activation.html) by populating the control container's user-data with a base64-encoded JSON block.

Reviewer: Same here.

Also I don't think we need the hyphen in "user-data" or "base64-encoded".  Editor @tjkirch might throw something at me... :)

Reviewer 2: The EC2 standard in prose is "user data", with a space and no hyphen.  "base64-encoded" should be hyphenated when used as an adjective like it is here.  :)

Author: Up to this point I've been using "user-data" to make it more clear that this is "host-containers.user-data" and not EC2's "user data". I take it I should just refer to both as user data going forward?

Reviewer 2: In the specific case referred to here, I'd say it's prose, since we're still in the "introduction", talking about the user data belonging to a container and not exactly describing how to set it with a setting that happens to be called `user-data`.  If you added the word "setting" afterward, then it'd be more specific, but I don't think we're to the "here's exactly how you do this" phase in the doc quite yet.

Down below, where it's saying the control container's "user-data setting" I'd go with the hyphen, plus backticks, to make it clear that now we're talking about a technical thing, a specific setting name, `user-data`.  (You also have the example TOML, which makes it clear that a hyphen is needed for that format.)

To use hybrid activations for managed instances you will want to generate a JSON-structure like this:

```
{

>   "ssm": {

Reviewer: Where would one find this information?  Can we point them to the top level docs for this?

    "activation-id": "foo",
    "activation-code": "bar",
    "region":"us-west-2"
  }
}
```

> Once you've created your JSON, you'll need to base64-encode it and set it as the value of the control host container's user-data setting in your [instance user data toml](https://github.com/bottlerocket-os/bottlerocket#using-user-data).

Reviewer: Same here - I don't think we need the hyphens.

You'll need Docker 17.06.2 or later, for multi-stage build support.
Then run `make`!

## Connecting to AWS Systems Manager (SSM)

Starting from v0.5.0, users have the option to pass in their own activation information for SSM.
This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.

Users can add their [own activations](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-managed-instance-activation.html) by populating the control container's user data with a base64-encoded JSON block.

To use hybrid activations for managed instances you will want to generate a JSON-structure like this:

```
{
  "ssm": {
    "activation-id": "foo",
    "activation-code": "bar",
    "region":"us-west-2"
  }
}
```

> Once you've created your JSON, you'll need to base64-encode it and set it as the value of the control host container's `user-data` setting in your [instance user data toml](https://github.com/bottlerocket-os/bottlerocket#using-user-data).

Reviewer 2: ```suggestion
Once you've created your JSON, you'll need to base64-encode it and put it in the control container's `user-data` setting in your [instance user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
For example:
```

Reviewer 2: That sentence is still complex, so I do think the friendly "For example:" afterward helps :)

You'll need Docker 17.06.2 or later, for multi-stage build support.
Then run `make`!

## Connecting to AWS Systems Manager (SSM)

Starting from v0.5.0, users have the option to pass in their own activation information for SSM.
This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.

Users can add their [own activations](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-managed-instance-activation.html) by populating the control container's user data with a base64-encoded JSON block.

To use hybrid activations for managed instances you will want to generate a JSON-structure like this:

```
{
  "ssm": {
    "activation-id": "foo",
    "activation-code": "bar",

>     "region":"us-west-2"

Reviewer 3: These names aren't consistent with the user data used in the admin container (`foo-bar` vs `foo_bar`). Was this intentional? As much as we can, I'd prefer that the "convention" stays consistent for users. This aligns with Bottlerocket's `user-data.toml` naming (which uses skewers) so one *could* argue that this one is the "right way" and the admin is in the wrong.. but we already have the admin change out, so I'm more inclined to align with that user data as this user data and the admin's, to me, roughly feels like it'll be treated as same namespace and with similar conventions by end users.

Author: Based on the other settings I see when I run `apiclient -u /settings` I would say that the admin container is the weird one, and that `foo-bar` is the "correct" convention, but I do not have strong feelings. I think the it might be worth getting some additional opinions. I think the admin container is the way it is because the authorized keys are typically stored in a file named `authorized_keys`, but that may have been a poor decision on my part. To make this conundrum even more fun, when you generate an activation token with `aws ssm create-activation` it returns JSON with a `FooBar` convention.

Author: After some offline discussion, we're going to keep the admin container convention and use `foo_bar` (snake_case) for the settings inside host-container user data in general. I'll add this change.

Reviewer 4: I'm in favor of sticking with the skewer (`foo-bar`) implementation here and adding a backlog item to fix the admin container to accept the skewer also.

  }
}
```

> Once you've created your JSON, you'll need to base64-encode it and put it in the control host container's `user-data` setting in your [instance user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
> 
> For example:

Reviewer 3: I took a stab at rewording this because it took a pass or two to full parse the sentence. Now that its split up I think this conveys the same and reads a bit easier (IMO) - what do you think about this instead?

```suggestion
Once you've created your JSON, you'll need to provide it via the control container's user data.
The host-container `user-data` must be base64-encoded and provided - in its encoded form - using your [instance's user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data):
```

Author: I think I prefer the original, but I appreciate that you're looking at the documentation 😃

```
[settings.host-containers.control]

> # ex: echo '{"ssm":{"activation-id":"foo","activation-code":"bar","region":"us-west-2"}}' | base64

Reviewer 5: On my system, I need `base64 -w0` to avoid newlines in the output.
```suggestion
# ex: echo '{"ssm":{"activation-id":"foo","activation-code":"bar","region":"us-west-2"}}' | base64 -w0
```

Author: I think we avoided it since not all systems may necessarily have `-w`. (MacOS, for example, does not)
